fuse_core/src/time.cpp

```diff
@@ -0,0 +1,136 @@
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+
+/*
+ * This file provides a sane configuration for a clock to manage order of events.
+ * This file exists to help with a migration from fuse_core::TimeStamp to rclcpp::time
+ * Priority is given to C++ std::chrono
+ * emphasis will be placed on interoperability with fuse_core::TimeStamp and rclcpp::time
+ *
+ * This will likely be extended to distinguish between:
+ *    times of events being optimised
+ *    events generated by the running of the optimiser (log throttling, callbacks etc)
+ *
+ */
+
+#include <fuse_core/time.h>
+#include <rclcpp/utilities.hpp>
+
+using chrono_ns_time_point_t =
+  std::chrono::time_point<std::chrono::system_clock, std::chrono::nanoseconds>;
+
+namespace fuse_core
+{
+TimeStamp::TimeStamp() :
+  rclcpp::Time(static_cast<int64_t>(0), RCL_CLOCK_UNINITIALIZED)
+{}
+
+TimeStamp::TimeStamp(const TimeStamp& rhs) = default;
+
+TimeStamp::TimeStamp(int32_t seconds, uint32_t nanoseconds, rcl_clock_type_t clock_type) :
+  rclcpp::Time(seconds, nanoseconds, clock_type)
+{}
+
+TimeStamp::TimeStamp(int64_t nanoseconds, rcl_clock_type_t clock_type) :
+  rclcpp::Time(nanoseconds, clock_type)
+{}
+
+TimeStamp::TimeStamp(const rclcpp::Time& time) :
+  rclcpp::Time(time)
+{}
+
+TimeStamp::TimeStamp(chrono_ns_time_point_t& stamp):
+  rclcpp::Time(stamp.time_since_epoch().count(), RCL_SYSTEM_TIME)
+{}
+
+rclcpp::Time TimeStamp::to_ros() const
+{
+  return rclcpp::Time(this->nanoseconds(), this->get_clock_type());
+}
+
+chrono_ns_time_point_t TimeStamp::to_chrono() const
+{
+  if (this->get_clock_type() != RCL_SYSTEM_TIME) {
+    throw std::runtime_error("Constructing system clock time_point from non-system clock");
+  }
+  return chrono_ns_time_point_t(
+    std::chrono::nanoseconds(this->nanoseconds())
+  );
+}
+
+
+// UTILITIES =======================================================================================
+std::ostream& operator<<(std::ostream& os, const fuse_core::TimeStamp& timestamp)
+{
+    os << timestamp.nanoseconds();
+    return os;
+}
+
+
+bool isValid(rclcpp::Clock::SharedPtr clock)
+{
+  switch (clock->get_clock_type()) {
+    case RCL_CLOCK_UNINITIALIZED:
+      return false;
+    case RCL_SYSTEM_TIME:
+      return true;
```

Comment:
The [design doc says](http://design.ros2.org/articles/clock_and_time.html) `A time value of zero should be considered an error meaning that time is uninitialized`, but it doesn't make it clear if that's for all clock types or just the ROS clock type.

`tf2` [considers zero on the system clock to be uninitialized](https://github.com/ros2/geometry2/blob/de08d177d424f0d73efd4ff5951fd1eaabc9638d/tf2/include/tf2/time.h#L45), so I'd recommend keeping that expectation here.

Reply:
If the method name is the same as from ROS 1, should we follow the logic from ROS 1 instead?

```
   bool Time::isValid()
   {
     return (!g_use_sim_time) || !g_sim_time.isZero();
   }
```

The isValid there checks for only time zero on simulation time (which I'd interpret as RCL_ROS_TIME and 0 (because it would be quite a roundabout route to trying to pull up the `use_sim_time` param..))

Reply:
My memory of ros1 time is pretty hazy, but I thought it ros1 started time at zero on all nodes when roscore started

ROS2 breaks some convenient assumptions by allowing nodes to start at any time, and typically wrapping various system clocks directly. if a wrapped clock produces a number, you can only assume that it's valid.

I think this switch statement is as sane as it gets

Reply:
Given that rcutils and rclcpp forbid negative time regardless of clock type https://github.com/locusrobotics/fuse/pull/283/files#r1010608091, I think `IsValid()` doesn't need to consider it. `IsValid` only needs to check time is greater than zero.

Reply:
Codified 0 time as invalid, duplicated the implementation in https://github.com/ros2/rclcpp/pull/2040 and used them https://github.com/locusrobotics/fuse/pull/283/commits/96b055546697e9c28ed6208383c4379c08fe2367